bgpd/bgp_flowspec_vty.c

```diff
@@ -0,0 +1,316 @@
+ */
+
+#include <zebra.h>
+#include "command.h"
+
+#include "bgpd/bgpd.h"
+#include "bgpd/bgp_table.h"
+#include "bgpd/bgp_attr.h"
+#include "bgpd/bgp_ecommunity.h"
+#include "bgpd/bgp_vty.h"
+#include "bgpd/bgp_route.h"
+#include "bgpd/bgp_aspath.h"
+#include "bgpd/bgp_flowspec.h"
+#include "bgpd/bgp_flowspec_util.h"
+#include "bgpd/bgp_flowspec_private.h"
+#include "bgpd/bgp_debug.h"
+
+/* Local Structures and variables declarations
+ * This code block hosts the struct declared that host the flowspec rules
+ * as well as some structure used to convert to stringx
+ */
+
+static const struct message bgp_flowspec_display_large[] = {
+	{FLOWSPEC_DEST_PREFIX, "Destination Address"},
+	{FLOWSPEC_SRC_PREFIX, "Source Address"},
+	{FLOWSPEC_IP_PROTOCOL, "IP Protocol"},
+	{FLOWSPEC_PORT, "Port"},
+	{FLOWSPEC_DEST_PORT, "Destination Port"},
+	{FLOWSPEC_SRC_PORT, "Source Port"},
+	{FLOWSPEC_ICMP_TYPE, "ICMP Type"},
+	{FLOWSPEC_ICMP_CODE, "ICMP Code"},
+	{FLOWSPEC_TCP_FLAGS, "TCP Flags"},
+	{FLOWSPEC_PKT_LEN, "Packet Length"},
+	{FLOWSPEC_DSCP, "DSCP field"},
+	{FLOWSPEC_FRAGMENT, "Packet Fragment"},
+	{0}
+};
+
+static const struct message bgp_flowspec_display_min[] = {
+	{FLOWSPEC_DEST_PREFIX, "to"},
+	{FLOWSPEC_SRC_PREFIX, "from"},
+	{FLOWSPEC_IP_PROTOCOL, "proto"},
+	{FLOWSPEC_PORT, "port"},
+	{FLOWSPEC_DEST_PORT, "dstp"},
+	{FLOWSPEC_SRC_PORT, "srcp"},
+	{FLOWSPEC_ICMP_TYPE, "type"},
+	{FLOWSPEC_ICMP_CODE, "code"},
+	{FLOWSPEC_TCP_FLAGS, "flags"},
+	{FLOWSPEC_PKT_LEN, "pktlen"},
+	{FLOWSPEC_DSCP, "dscp"},
+	{FLOWSPEC_FRAGMENT, "pktfrag"},
+	{0}
+};
+
+#define	FS_STRING_UPDATE(count, ptr, format) do {			\
+		if (((format) == NLRI_STRING_FORMAT_DEBUG) && (count)) {\
+			(ptr) += sprintf((ptr), ", "); \
+		} else if (((format) == NLRI_STRING_FORMAT_MIN) && (count)) {\
+			(ptr) += sprintf((ptr), " "); \
+		} \
+		count++;			\
+	} while (0)
+
+/* Parse FLOWSPEC NLRI*/
+void bgp_fs_nlri_get_string(unsigned char *nlri_content, size_t len,
+			    char *return_string, int format)
+{
+	u_int32_t offset = 0;
+	int type;
+	int ret = 0, error = 0;
+	char *ptr = return_string;
+	char local_string[BGP_FLOWSPEC_STRING_DISPLAY_MAX];
+	int count = 0;
+	char extra[2] = "";
+	char pre_extra[2] = "";
+	const struct message *bgp_flowspec_display;
+
+	if (format == NLRI_STRING_FORMAT_LARGE) {
+		sprintf(pre_extra, "\t");
+		sprintf(extra, "\n");
+		bgp_flowspec_display = bgp_flowspec_display_large;
+	} else
+		bgp_flowspec_display = bgp_flowspec_display_min;
+	error = 0;
+	while (offset < len-1 && error >= 0) {
+		type = nlri_content[offset];
+		offset++;
+		switch (type) {
+		case 1:
+		case 2:
+			ret = bgp_flowspec_ip_address(
+						BGP_FLOWSPEC_RETURN_STRING,
+						nlri_content+offset,
+						len - offset,
+						local_string, &error);
+			if (ret <= 0)
+				break;
+			FS_STRING_UPDATE(count, ptr, format);
+			ptr += sprintf(ptr, "%s%s %s%s", pre_extra,
+				     lookup_msg(bgp_flowspec_display, type, ""),
+				     local_string, extra);
+			break;
+		case 3:
```

Comment:
Please, use switch / case range.
  https://gcc.gnu.org/onlinedocs/gcc/Case-Ranges.html

Reply:
Actually, values are defined by flowspec, and I will add the define names instead of the numbers.
For the range case, for the same reason, I am not sure, since I can have values that are not following.